[ModuleInterface] Support classes with missing designated inits

harlanhaskins opened the pull request:
When a superclass in one module has a designated init that is non-public (or
@usableFromInline), and another class attempts to subclass that class from a
different module, we would have enough information in a binary module to
know that the class can't be subclassed because it's missing designated
initializers.

But if we're in a module interface, there's no indication that there was
an un-printed initializer. Introduce a new attribute
@_hasMissingDesignatedInits on classes that have non-public or usable
from inline designated initializers, and ensure that the bit is set when
we parse one.

Also, if a class inherits from something that has missing designated inits, but it overrides all designated initializers anyway, it should be allowed to inherit convenience inits _even if they're missing from the interface_. As such, add another attribute `@_inheritsSuperclassInitializers`, and promote the existing bit to an attribute. This'll ensure that we follow initializer inheritance the same way we would with bits in the serialized module.

Fixes rdar://51249311

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -3778,10 +3777,17 @@ class ClassDecl final : public NominalTypeDecl {
+    unsigned HasMissingDesignatedInitializers : 1;
+
+    /// Whether we have computed the above field or not.
+    unsigned HasMissingDesignatedInitializersComputed : 1;
```
I think it's fine to not have a cached bit for this at all. Just checking the attribute is fine.

jrose-apple commented on include/swift/AST/Decl.h:
```diff
@@ -3900,8 +3906,16 @@ class ClassDecl final : public NominalTypeDecl {
+    LazySemanticInfo.HasMissingDesignatedInitializersComputed = true;
+
+    // Add an attribute for printing
+    if (value && !getAttrs().hasAttribute<HasMissingDesignatedInitsAttr>()) {
```
This doesn't handle *removing* the attribute if it's set to `false`. There's no real reason to implement that, but maybe assert that it's not supported?

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2379,9 +2379,43 @@ void PrintAST::visitStructDecl(StructDecl *decl) {
+static bool hasInaccessibleDesignatedInit(ClassDecl *decl) {
+  // We only care for classes that show up in module interfaces.
+  if (!isPublicOrUsableFromInline(decl))
+    return false;
```
I don't think this is necessary to check; if we're printing the class, we already know we want this.

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2379,9 +2379,43 @@ void PrintAST::visitStructDecl(StructDecl *decl) {
+    auto init = dyn_cast<ConstructorDecl>(member);
+    if (!init->isDesignatedInit())
+      continue;
+    if (!isPublicOrUsableFromInline(init))
```
…and then to make this fit ASTPrinter better, you can use `shouldPrint` instead of hardcoding `isPublicOrUsableFromInline`.

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2379,9 +2379,43 @@ void PrintAST::visitStructDecl(StructDecl *decl) {
+
+  for (auto member : members) {
+    assert(isa<ConstructorDecl>(member));
+    auto init = dyn_cast<ConstructorDecl>(member);
```
These two lines are just `cast<ConstructorDecl>(member)`.

jrose-apple reviewed (commented):
Can you add tests for the internal subclass case? I think we haven't completely solved the problem without saying "but *this* class *does* override all the designated initializers, even though you can't see them".

slavapestov commented:
This should replace ClassDecl::inheritsSuperclassInitializers() and setInheritsSuperclassInitializers() too, together with the existing code for computing that in TypeChecker::addImplicitConstructors().

harlanhaskins commented:
I'd really like, at some point, to pull out this logic from addImplicitConstructors, but it's really tightly woven with the logic for diagnosing missing overrides and creating implicit overrides.

harlanhaskins commented:
@swift-ci please smoke test

jrose-apple commented on include/swift/AST/Attr.def:
```diff
@@ -414,6 +414,15 @@ DECL_ATTR(_projectedValueProperty, ProjectedValueProperty,
+  90)
+
+SIMPLE_DECL_ATTR(_inheritsSuperclassInitializers,
+  InheritsSuperclassInitializers,
```
This should really be `InheritsSuperclassConvenienceInitializers`, or just `InheritsConvenienceInitializers`. (We already print inherited designated initializers, right?)

jrose-apple commented on test/ParseableInterface/inherits-superclass-initializers-client.swift:
```diff
@@ -0,0 +1,27 @@
+
+import Module
+
+test()
```
It's also important that `SubSub()` is visible *here* as well.

jrose-apple commented on test/ParseableInterface/inherits-superclass-initializers-client.swift:
```diff
@@ -0,0 +1,27 @@
+// Compile the imported module to a .swiftinterface and ensure the convenience
+// init cannot be called.
```
This comment doesn't match the test.

jrose-apple commented on test/ParseableInterface/inherits-superclass-initializers-client.swift:
```diff
@@ -0,0 +1,27 @@
+// R/UN: rm %t/Module.swiftmodule
+// R/UN: %target-build-swift %s -I %t -L %t -lModule -o %t/main
+// R/UN: %target-codesign %t/main %t/libModule.dylib
+// R/UN: %target-run %t/main %t/libModule.dylib | %FileCheck %s
```
What happened to these RUN lines?

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
LLDB failure is unrelated.

@swift-ci please test Linux

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

jrose-apple commented on lib/AST/Decl.cpp:
```diff
@@ -3900,12 +3898,7 @@ GetDestructorRequest::evaluate(Evaluator &evaluator, ClassDecl *CD) const {
-  (void)mutableThis->lookupDirect(DeclBaseName::createConstructor(),
-                                  flags);
-  return Bits.ClassDecl.HasMissingDesignatedInitializers;
+  return getAttrs().hasAttribute<HasMissingDesignatedInitsAttr>();
```
This still needs to do the `lookupDirect`, because forcing the inits to be loaded is what determines if any are missing in a compiled module (and soon the importer too).

jrose-apple commented:
The awful case we talked about elsewhere:

```swift
// Library.swift
public /* not open */ class Base {
  public init() {}
  public convenience init(foo: Int) {
    self.init()
    print(foo)
  }
}

public class Sub: Base {
  public override init() {}
}
```

```swift
// Client.swift
import Library
extension Base {
  public convenience init(bar: Int) {
    self.init()
    print(bar)
  }
}

let sub1 = Sub(foo: 42)
let sub2 = Sub(bar: 42)
```

If Library adds a new designated init to Base (public or non-public), then Sub's "inherits convenience initializers" status changes, which is…unfortunate. We can check for this kind of thing in the API checker (I'll file a bug), but it's too bad that it's not opt-in the way nearly everything else exposed to the public is.

nkcsgexi commented:
Just to consolidate the related impact on ABI/API checker, should we diagnose the adding of `@_hasMissingDesignatedInits` and the removing of `@_inheritsSuperclassInitializers` after this change?

jrose-apple commented:
Adding `@_hasMissingDesignatedInits` is breaking only if the class is `open`; removing `@_inheritsSuperclassInitializers` is breaking always.

jrose-apple commented:
Adding a public designated initializer is also breaking if the class is `open`.

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

jrose-apple commented on include/swift/AST/Attr.def:
```diff
@@ -497,6 +497,17 @@ DECL_ATTR(_projectedValueProperty, ProjectedValueProperty,
 
+SIMPLE_DECL_ATTR(_hasMissingDesignatedInits, HasMissingDesignatedInits,
+  OnClass | UserInaccessible | APIStableToAdd | ABIStableToAdd |
+  APIBreakingToRemove | ABIBreakingToRemove,
```
I don't think it's APIBreakingToRemove, is it? ABI, maybe, since new things that depend on this won't backwards deploy, but…

jrose-apple commented on lib/AST/ASTPrinter.cpp:
```diff
@@ -2389,6 +2389,32 @@ void PrintAST::visitStructDecl(StructDecl *decl) {
+        continue;
+      if (!shouldPrint(init))
+        return true;
+    }
```
Would this be clearer with `llvm::any_of`? You might not even need the lambda then.

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test source compatibility

harlanhaskins commented:
@swift-ci please clean test

harlanhaskins commented:
@swift-ci please test source compatibility

harlanhaskins commented:
@swift-ci please smoke test

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -3787,22 +3783,62 @@ class ClassDecl final : public NominalTypeDecl {
+      (void)mutableThis->lookupDirect(DeclBaseName::createConstructor());
+    }
+
+    bool hasAttr = getAttrs().hasAttribute<HasMissingDesignatedInitsAttr>();
```
Checking for the attribute should be part of the request's evaluate() method -- it can short-circuit the actual computation there.

slavapestov commented on include/swift/AST/Decl.h:
```diff
@@ -3787,22 +3783,62 @@ class ClassDecl final : public NominalTypeDecl {
+    auto existing = getAttrs()
+        .getAttribute<HasMissingDesignatedInitsAttr>();
+    if (value && !existing) {
+      auto attr = new (getASTContext())
```
Lets not synthesize new attributes here. Instead, serialize the bit directly.

harlanhaskins replied:
We need to print the attribute in the module interface

slavapestov replied:
Can you handle that entirely in the ASTPrinter or would it be too messy? We have similar issues with @objc IIRC.

harlanhaskins replied:
That's a good point, I can move back to this being a bit and then the attribute can just override the bit.

harlanhaskins replied:
Ah, no, we want to diagnose whether this attribute was added or removed, because it changes ABI. I'd rather keep this as an attribute.

harlanhaskins replied:
Though, I suppose those can just become semantic checks in the api digester

harlanhaskins committed:
[Serialization] Serialize hasMissingDesignatedInitializers

Since this is going to be something modules tell clients, rather than something clients discover about modules, serialize it.

harlanhaskins committed:
[Sema] Requestify hasMissingDesignatedInitializers

We’re going to start serializing this for public types that have non-public-or-@usableFromInline initializers, so turn it into a request that we can query and cache it in the existing bit.

harlanhaskins committed:
[ModuleInterface] Add printing for new attributes

Specially print @_hasMissingDesignatedInitializers and @_inheritsConvenienceInitializers in module interfaces

Fixes rdar://51249311

harlanhaskins committed:
[api-digester] Teach the api-digester about hasMissingDesignatedInitializers

Because we won’t be serializing this attribute, add custom diagnostics for the cases where:

- We add @_hasMissingDesignatedInits to an open class, which means subclasses won’t be able to inherit its inits
- We remove @_inheritsConvenienceInitializers, which means APIs are removed

harlanhaskins commented:
@swift-ci please smoke test macOS

harlanhaskins commented:
@nkcsgexi Can you look at the api-digester changes here and tell me if they make sense?

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please clean test Linux

harlanhaskins commented:
@swift-ci please test

harlanhaskins commented:
@swift-ci please test

CodaFi commented:
Shepherded in #28629
